server/src/main/java/com/tencent/rss/server/buffer/ShuffleBuffer.java

```diff
@@ -0,0 +1,374 @@
+/*
+ * Tencent is pleased to support the open source community by making
+ * Firestorm-Spark remote shuffle server available. 
+ *
+ * Copyright (C) 2021 THL A29 Limited, a Tencent company.  All rights reserved. 
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * https://opensource.org/licenses/Apache-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package com.tencent.rss.server.buffer;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.tencent.rss.common.BufferSegment;
+import com.tencent.rss.common.ShuffleDataResult;
+import com.tencent.rss.common.ShufflePartitionedBlock;
+import com.tencent.rss.common.ShufflePartitionedData;
+import com.tencent.rss.common.util.Constants;
+import com.tencent.rss.server.ShuffleDataFlushEvent;
+import com.tencent.rss.server.ShuffleFlushManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+
+public class ShuffleBuffer {
```

Comment:
Could we break the list of `ShufflePartitionBlocks` into index buffer and data buffer since we support reading directly from memory?

Reply:
there will be a lot of System.arrayCopy which is an expensive call

Reply:
OK